### PR TITLE
REVSDL-2384 RegisterService request implementation

### PR DESCRIPTION
**Pull request by thinhts**
Please review
@DKlimenko 
@bonewell 
@KKarlash 
@VladSemenyuk 
@giangtqh

### CODE REVIEW

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 1]
```diff
@@ -40,14 +41,14 @@ namespace commands {
 CREATE_LOGGERPTR_GLOBAL(logger_, "VRCooperation")
 
 BaseCommandRequest::BaseCommandRequest(
-  const application_manager::MessagePtr& message)
-  : message_(message) {
+    VRModule* vr_module, const application_manager::MessagePtr& message)
+    : message_(message),
+      vr_module_(vr_module){
```
Comment (bonewell) [inline src/components/vr_cooperation/src/base_command_request.cc]: Space between ) and { will be good.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 2]
```diff
@@ -38,19 +38,24 @@
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: `#include <string>`

File: src/components/vr_cooperation/include/vr_cooperation/vr_module.h [hunk 3]
```diff
@@ -113,6 +113,11 @@ class VRModule : public functional_modules::GenericModule,
    */
   void ReceiveMessageFromMobile();
 
+  /**
+   * @brief Send notification to HMI
+   */
+  void SendNotificationToHMI(application_manager::MessagePtr msg);
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module.h]: description for param

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 4]
```diff
@@ -31,6 +31,7 @@
  */
 
 #include "vr_cooperation/commands/base_command_request.h"
+#include "vr_cooperation/vr_module.h"
 #include "utils/logger.h"
```
Comment (giangtqh) [inline src/components/vr_cooperation/src/base_command_request.cc]: alphabet include order

File: src/components/vr_cooperation/include/vr_cooperation/commands/register_service_request.h [hunk 5]
```diff
@@ -64,7 +65,8 @@ class RegisterServiceRequest : public BaseCommandRequest {
   /**
    * @brief This method will be called whenever new event received
    */
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/register_service_request.h]: add description for params

**Comment by giangtqh** [comment]
reviewed

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 6]
```diff
@@ -71,6 +73,36 @@ void BaseCommandRequest::SendRequest() {
   LOG4CXX_AUTO_TRACE(logger_);
 }
 
+void BaseCommandRequest::SendNotification(const char* function_id,
+                                          const Json::Value& message_params) {
+  Json::Value msg;
```
Comment (giangtqh) [inline src/components/vr_cooperation/src/base_command_request.cc]: LOG4CXX_AUTO_TRACE(logger_);

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 7]
```diff
@@ -89,6 +94,14 @@ class BaseCommandRequest : public Command {
   void SendRequest();
 
   /**
+   * @brief send request to HMI
```
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: send request?

File: src/components/vr_cooperation/src/register_service_request.cc [hunk 8]
```diff
@@ -40,18 +41,26 @@ namespace commands {
...
 RegisterServiceRequest::~RegisterServiceRequest() {
 }
 
 void RegisterServiceRequest::Execute() {
   LOG4CXX_AUTO_TRACE(logger_);
+  SendNotification(functional_modules::hmi_api::on_register_service,
+                   MessageHelper::StringToValue(message_->json_message()));
+  std::string result_code;
+  std::string info;
+  bool success = true;
```
Comment (KKarlash) [inline src/components/vr_cooperation/src/register_service_request.cc]: should be depend on result code

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 9]
```diff
@@ -38,19 +38,24 @@
...
 namespace commands {
 
 /**
  * @brief Base command class for requests
  */
-class BaseCommandRequest : public Command {
-  // TODO(Thinh) inherit from EventEngine
+class BaseCommandRequest : public Command,
+  public event_engine::EventObserver<
+  application_manager::MessagePtr, std::string> {
```
Comment (DKlimenko) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: std::string could be const reference

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 10]
```diff
@@ -71,6 +73,36 @@ void BaseCommandRequest::SendRequest() {
...
+  if (!message_params.isNull()) {
+    msg[kParams] = message_params;
+  }
+
+  msg[kParams][json_keys::kAppId] = app_->hmi_app_id();
+
+  Json::FastWriter writer;
+
+  application_manager::MessagePtr message_to_send(
+      new application_manager::Message(
```
Comment (DKlimenko) [inline src/components/vr_cooperation/src/base_command_request.cc]: no::throw

File: src/components/vr_cooperation/src/base_command_request.cc (hunk 1 above)
Comment (DKlimenko) [inline src/components/vr_cooperation/src/base_command_request.cc]: +1

**Comment by DKlimenko** [comment]
reviewed

**Comment by thinhts** [comment]
I fixed all comment.Please review
@DKlimenko 
@bonewell 
@KKarlash 
@VladSemenyuk 
@giangtqh

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 11]
```diff
@@ -47,7 +48,7 @@ namespace commands {
  */
 class BaseCommandRequest : public Command,
   public event_engine::EventObserver<
-  application_manager::MessagePtr, std::string> {
+  application_manager::MessagePtr, const std::string&> {
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: I'm not sure about using const& for string. Who creates and deletes these strings?

**Comment by thinhts** [comment]
rebase and update code
@DKlimenko 
@bonewell 
@KKarlash 
@VladSemenyuk 
@giangtqh

**Comment by bonewell** [comment]
Reviewed.

**Comment by bonewell** [comment]
Reviewed.

**Comment by DKlimenko** [comment]
reviewed

**Comment by VladSemenyuk** [comment]
reviewed

**Pull request closed without merging**